### ISSUE TITLE
Cashu spec related questions

### ISSUE THREAD

**Issue by misovan** [cashubtc/nuts#30]
Hi, first let me thank all the folks working on Cashu. The more I think and dig into it, the more inspirational it looks.

I recently started working on a mobile wallet with the aim of exploring the design space to make eCash work in a more comprehensible way for the average user accustomed to their bank's mobile app. The goal is to enable seamless payments and receipt of funds without the need to copy payment request formats, deep links, or QR codes, etc. I'm close to having an MVP wallet ready and would appreciate some explanations/opinions on a few Cashu-related concepts:

Is there a strong need for the current token format that supports encoding tokens issued by multiple mints at once?
Reasoning: At first glance, it's not an intuitive concept and it makes reasoning, design, and implementation of sensitive code more complex. As far as I'm aware, current wallets only send single mint tokens. However, to comply with the specification, the wallet needs to be ready to accept tokens from multiple mints. It's similar to a payment funded from multiple banks/accounts at once, which is easily solved by a sequence of transactions. If there is a longer-term strategy underlying this, please explain.

The above issue manifests itself in the Cashu-ts library where there's a situation where a single CashuWallet class instance is constructed with a single mint URL but then implements the receive function from multiple mints. If anyone knows of any plans on how this implementation will conceptually evolve, I would gladly structure my code in that direction.

Naming conventions: I understand that the protocol is new and not finished, but I've been confused many times about how many different things are called a "token" in the specification, documentation, and the code itself. It ranges from encoded serialized tokens to proof collections indexed by mints and the proofs themselves. Would it be possible to propose concise naming conventions? The TypeScript types in Cashu-ts are perhaps the clearest ones.

I understand that it is very difficult to change widely implemented protocols later, even in networks without proof of work. However, every time I dig through those V3 token hierarchies, the idea comes to my mind to add the damn mintUrl to the proof, attach a memo, and forget about them once and for all. All critical code would then work on a simple array of flat objects. :)

The last conceptual point is a fundamental naming convention - how should Cashu tokens be represented in the wallet? The current standard is to represent them as "sats" (as they are hopefully backed by real sats within the mint). However, this raises two questions: Is it not misleading? Would it be more transparent to adopt a convention similar to stablecoins that transparently communicate they are not the same as what they are backed with? As a follow-up question, should we not consider making the protocol ready to issue tokens backed by something else that is digitally transferable (ideally, but not necessarily only through Lightning)? Stablecoins might be a good candidate.

Thanks to everybody reading this and for all your opinions.
MV

**Comment by callebtc** [comment]
> Is there a strong need for the current token format that supports encoding tokens issued by multiple mints at once?

Yes, as you've mentioned correctly, we wanted to leave room for potentially encoding tokens from multiple mints in the same token. This would allow a "multimint" payment in the future.

> CashuWallet class instance is constructed with a single mint URL but then implements the receive function from multiple mints.

I guess the reason here is that the lib has implemented (only) parts of the idea above: mint, melt, send from a single mint (hence the initialization with a single mint URL), but your wallet wants to be able to receive from multiple mints (hence the multimint receive function). You will find a similar design in the Nutshell wallet (Python).

> Naming conventions: I understand that the protocol is new and not finished, but I've been confused many times about how many different things are called a "token" in the specification, documentation, and the code itself.

You are correct, the naming is hard and we're trying to keep it so that it can't be easily confused. We usually refer to a non-serialized proof as `Proof` and we call a serialized proof that includes mint information a `Token`. Therefore, `Token` should always refer to a serialized object that can be passed around from wallet to wallet. I hope that we are consistent everywhere, if you find any cases where we aren't please help clean up mistakes. 

Note: Inside the serialization format `TokenV3` is a nested `token` field that holds all the actual token-related information and a `memo` field that can be used between wallets. This is one point of confusion, I guess but I hope it's not too bad. 

> the idea comes to my mind to add the damn mintUrl to the proof, attach a memo to the list of proofs to be sent, and forget about them once and for all.

This is exactly what `TokenV3` achieves :)

> Would it be possible to propose concise naming conventions?

In principle yes, but we don't optimize anything specifically for efficiency yet. If you find clear naming mistakes, please propose changes! To make things more efficient though, it might be better to come up with a new (binary) `TokenV4` though, which we postponed until needed.

> The last conceptual point is a fundamental naming convention - how should Cashu tokens be represented in the wallet?

We all use "sat" but we could also call it differently. In any case, it is exchanged 1-to-1 with real sats on LN, so the question is rather how much explicitness is necessary for a user to get any value from that. Wallet of Satoshi could also call their denomination "wSats" but I don't think that adds any value but rather would confuse a lot of users.

Sorry for the late reply, I was gone for a while. Do you mind sharing more details of your project? Do you have a repo yet? It sounds very interesting! Also, please join our dev channel on Telegram!

**Comment by callebtc** [comment]
> I understand that it is very difficult to change widely implemented protocols later, even in networks without proof of work. However, every time I dig through those V3 token hierarchies (Token, Token.token, TokenEntry, Proof), the idea comes to my mind to add the damn mintUrl to the proof, attach a memo to the list of proofs to be sent, and forget about them once and for all. All critical code would then work on a simple array of flat objects. :) 

I guess there are two extremes of this approach: either you put all information into every Proof, or you use some kind of tree structure to save data. In TokenV3, Keyset IDs for example, repeat very often. Adding the Mint URL as well to each proof would significantly contribute to the size of the serialized token.

If there is a better way to group things, please feel free to suggest one that is easy to use for devs and has a small footprint. My gut feeling is we can use this format for a while but then transition to a bech32 encoded TLV binary format and focus on keeping the size low.

In Nutshell, I immediately parse and unpack a serialized token into a simple array of Proofs and use it for all internal logic the way you described it. 

Btw, would you DM me on any of the socials? Peace!

**Comment by misovan** [comment]
Fully agree, it does not make sense to repeat information in serialized (sendable) proofs. I'd suggest to keep current format and move on to define binary one. 

From wallet dev perpspective, I as well ended up to work with proofs only, I've just extended the internal Proof type by needed properties (txId, mintUrl). On send, I just strip those properties and create sendable token according to protocol. So it can be dealt with in reasonably clean way by decoupling the internal wallet format (that might be completely outside of protocol scope) from serialized sendable ones. 

On the other side, sendable tokens become quite regularly above soft 1000 char QR code limits, and inability to scan them is an UX nightmare.

Regarding dm, I tried to send you twitter message from my new project's account, but it might have been blocked by twitter, idk. 

You should now have access to minibits-cash/minibits_wallet github repo with an alpha version of the wallet code and quite detailed redme that I keep private for now to let it be checked by few more people and do more testing. If you would have few minutes I would appreciate you to go through the readme. I don' use telegram but if this is where all cashu folks are I'll hold my breath and install it.
Thnx for thoughtful discussion
MV

**Comment by callebtc** [comment]
Thank you, I will recheck my DMs. I've already looked through the minibits repo and I'm simply blown away. Really incredible work. Will fix my `adb` and try to load it onto my Android as soon as I can.

If you don't want to use Telegram, you can also join this Matrix channel that has a Telegram bridge to the main room: https://matrix.to/#/%23cashu:matrix.org
